flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java

```diff
@@ -708,6 +708,20 @@ public class TaskManagerOptions {
                             "Time we wait for the timers in milliseconds to finish all pending timer threads"
                                     + " when the stream task is cancelled.");
 
+    public static final ConfigOption<TaskManagerLoadBalanceMode> TASK_MANAGER_LOAD_BALANCE_MODE =
```

Comment:
It likes other options, it's better to add some comments, and     `@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)`?

Or should we add `@Documentation.Section` and deprecate `slot-spread-out`  after all PRs is merged?

Reply:
Thank you very much for the precise comments !

I'd prefer the second and after a little adjustment as follows:
1.  Add `@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)`
2. Temporarily comment out 'SLOT' mode in the `TaskManagerLoadBalanceMode`
3.  Add the description about the configuration item like follows:
      ```
           The load balance mode of taskmanager when processing scheduling tasks.

           'NONE' means that the scheduler does not consider any dimensional balance when scheduling tasks.
           'TASKS' means that the scheduler prioritizes ensuring that the number of tasks on each TM is balanced when scheduling tasks.

           Please view FLIP-370 for more details.
      ```
4. After all PRs merged, we'll add the 'SLOTS' mode back, supplement the description and deprecate slot-spread-out. the added description would like follows:
      ```
          'SLOTS' indicates that the scheduler prioritizes and balances the use of each TM's slot when scheduling tasks.
      ```

Please let me know what's your opinion~

Reply:
Thanks @RocMarshal  for the analysis!

How about creating a JIRA to add `taskmanager.load-balance.mode` and deprecate `cluster.evenly-spread-out-slots` first?
- It don't introduce any feature, just using the new option and deprecate old option.
- It should be the first JIRA, and `TaskManagerLoadBalanceMode` just have 2 enums: `None` and `Slots`.
- In the second JIRA(maybe this PR), you can start your feature: adding the `Tasks`, and supports it.

I think this process is clearer and easier than the process you proposed, and it don't effect users side.

--------------------------

Besides how to implement this, I have a question about compatibility of `taskmanager.load-balance.mode` and `cluster.evenly-spread-out-slots`.

As I understand :
- `taskmanager.load-balance.mode : Slots`  == `cluster.evenly-spread-out-slots: true`
- `taskmanager.load-balance.mode : None`  == `cluster.evenly-spread-out-slots: false` (By default.)

After this FLIP is merged, if a job with `cluster.evenly-spread-out-slots: true`, and didn't set the `taskmanager.load-balance.mode`. Should we change the default value from `None` to `Slots`?

- If yes, we can support old users directly.
- If no, users must set `taskmanager.load-balance.mode : Slots` manually if they want to the `evenly-spread-out-slots` feature.

Reply:
hi, @1996fanrui Thanks for the idea. There's nothing better!

>After this FLIP is merged, if a job with cluster.evenly-spread-out-slots: true, and didn't set the taskmanager.load-
>balance.mode. Should we change the default value from None to Slots?

Yes,  I'd  prefer to be compatible with old configuration item.

- This can provide users with a smooth over experience
- Removing the old configuration item requires a process for evolution.  For example, after one or two editions, we will completely discard the old configuration item.

I'd create this JIRA ticket and advance the related work if there's nothing ambiguous.

Reply:
Please go ahead for the first JIRA, please cc me after the PR is ready, thanks~